latency test enhancement - part 1

min-xu-et opened the pull request:
This is part 1 of change to make latency test generate graph.

* this makes the batch_size arg a list, but user can continue use it the same way as before when specifying a single value
* this optionally dumps the results in a json file. example:

```
$ cat out.json
[
  {
    "batch_size": 1,
    "output_len": 256,
    "prefill_latency": 0.12703561782836914,
    "prefill_throughput": 4030.3657253962833,
    "avg_decode_latency": 0.05401019565761089,
    "avg_decode_throughput": 18.515022725326567,
    "total_latency": 13.953645706176758,
    "total_throughput": 55.03937939745991
  },
  {
    "batch_size": 1,
    "output_len": 256,
    "prefill_latency": 0.137282133102417,
    "prefill_throughput": 3729.545778677777,
    "avg_decode_latency": 0.054004695266485214,
    "avg_decode_throughput": 18.516908484818174,
    "total_latency": 13.962484121322632,
    "total_throughput": 55.00453882895798
  }
]
```

min-xu-et committed:
latency test enhancement - part 1

min-xu-et committed:
Merge remote-tracking branch 'upstream/main' into benchy

merrymercy commented on python/sglang/bench_latency.py:
```diff
@@ -218,15 +234,25 @@ def latency_test(
     )
 
+    # To make this PR easier to review, for now, only do the first element in batch_size tuple.
+    bench_args.batch_size = bench_args.batch_size[0]
```
Remove this line or do you want to do it in the follow up PR?

min-xu-et replied:
That's my thought to make the review easier. If you prefer, I can include in this PR another commit for the part 2 as well. For part 2, I plan to move the nested function out and call them in a loop.

min-xu-et replied:
please unresolve this if you like to combine step 1 and step 2 in this PR.

merrymercy commented on python/sglang/bench_latency.py:
```diff
@@ -267,13 +297,25 @@ def run_once(output_len):
+        if os.path.isfile(bench_args.result_filename):
+            data = json.loads(open(bench_args.result_filename, "r").read())
+        data += ret
+        open(bench_args.result_filename, "w").write(json.dumps(data, indent=2))
```
Can we use jsonl format with append mode?

min-xu-et replied:
Great idea. Will do.

min-xu-et replied:
resolved in the new commit https://github.com/sgl-project/sglang/pull/909/commits/f52110d52b4b96ced2dbd25dadc8c90d99af5ad6

min-xu-et committed:
use jsonl format

- testing output

```
In [3]: with open("out.jsonl", "r") as f:
   ...:     for l in f.readlines():
   ...:         print(json.loads(l))
   ...:         print("-"*80)
   ...:
{'batch_size': 1, 'output_len': 256, 'prefill_latency': 0.13420748710632324, 'prefill_throughput': 3814.9883515394195, 'avg_decode_latency': 0.053959593176841736, 'avg_decode_throughput': 18.532385830313817, 'total_latency': 13.947863340377808, 'total_throughput': 55.06219707334737}
--------------------------------------------------------------------------------
{'batch_size': 1, 'output_len': 256, 'prefill_latency': 0.13029885292053223, 'prefill_throughput': 3929.4282990523557, 'avg_decode_latency': 0.05398500803858042, 'avg_decode_throughput': 18.523661222488833, 'total_latency': 13.95046091079712, 'total_throughput': 55.051944513575}
--------------------------------------------------------------------------------
{'batch_size': 1, 'output_len': 256, 'prefill_latency': 0.13561320304870605, 'prefill_throughput': 3775.4436034971686, 'avg_decode_latency': 0.05398557987064123, 'avg_decode_throughput': 18.52346501410511, 'total_latency': 13.955921649932861, 'total_throughput': 55.03040352793143}
--------------------------------------------------------------------------------
```

min-xu-et committed:
use jsonlines library

- simpler code
